src/proxy.rs

```diff
@@ -124,13 +123,36 @@ impl<S: IntoUrl> IntoProxyScheme for S {
         let url = match self.as_str().into_url() {
             Ok(ok) => ok,
             Err(e) => {
-                // the issue could have been caused by a missing scheme, so we try adding http://
-                format!("http://{}", self.as_str())
-                    .into_url()
-                    .map_err(|_| {
+                let mut presumed_to_have_schema = true;
+                let mut source = e.source();
+                while let Some(err) = source {
+                    if let Some(parse_error) = err.downcast_ref::<url::ParseError>() {
+                        match parse_error {
+                            url::ParseError::RelativeUrlWithoutBase => {
+                                presumed_to_have_schema = false;
+                                break;
+                            }
+                            _ => {}
+                        }
+                    } else {
+                        let text = format!("{}", err);
+                        if text == "URL scheme is not allowed" {
```

Comment:
Matching on the `Display` output of an error makes me nervous. In pretty much all error types, the output is not considered stable. Is there no other way to check? Is this check important?

Reply:
Yeah.  That's not my proudest moment.

The check is important.  The `domain:port` case ends up [here](https://github.com/seanmonstar/reqwest/blob/master/src/error.rs#L268).  The string comparison was the easiest way to check for that error.  I think a better option would be to carry a typed thing instead of a string.  There's some risk but the reward is a better solution.  I'll poke at it for a bit.

Reply:
Oh, if that error is owned by the reqwest crate, we can just make it a private unit struct, instead of a string, and then check via downcasting.

Reply:
Yup.  Just pushed the update.  Windows looks good.  Checking Linux.  Running cargo fmt.  All good!